src/grpc/stream_infer_handler.cc

```diff
@@ -130,362 +130,424 @@ ModelStreamInferHandler::StartNewRequest()
+  // and writes to complete.
+  if (!state->context_->PopCompletedResponse(state)) {
+    LOG_ERROR << "Unexpected response for " << Name() << ", rpc_ok=" << rpc_ok
+              << ", context " << state->context_->unique_id_ << ", "
+              << state->unique_id_ << " step " << state->step_;
+    state->context_->finish_ok_ = false;
+  }
+
+  // Write the next response if it is ready...
+  state->context_->WriteResponseIfReady(nullptr);
+
+  // The response for the request has been written completely.
+  // The counter can be safely decremented.
+  state->context_->DecrementRequestCounter();
+  finished = Finish(state);
+  return finished;
+}
+
+bool
+ModelStreamInferHandler::RequestWriteReadyStep(InferHandler::State* state)
+{
+  if (state->is_decoupled_) {
+    return RequestWriteReadyStepDecoupled(state);
+  }
+  // Currently don't expect WRITEREADY state for non-decoupled model, as other
+  // other steps will call WriteResponseIfReady and quickly transition to
+  // WRITTEN state.
+  return false;
+}
+
+bool
+ModelStreamInferHandler::RequestWriteReadyStepDecoupled(
+    InferHandler::State* state)
+{
+  bool finished = false;
+  if (state->delay_response_ms_ != 0) {
+    // Will delay the write of the response by the specified time.
+    // This can be used to test the flow where there are other
+    // responses available to be written.
+    LOG_INFO << "Delaying the write of the response by "
+             << state->delay_response_ms_ << " ms...";
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(state->delay_response_ms_));
+  }
+
+  // Finish the state if all the transactions associated with
+  // the state have completed.
+  if (state->IsComplete()) {
+    state->context_->DecrementRequestCounter();
+    finished = Finish(state);
+  } else {
+    // GRPC doesn't allow to issue another write till
+    // the notification from previous write has been
+    // delivered. If there is an ongoing write then
+    // defer writing and place the task at the back
+    // of the completion queue to be taken up later.
+    if (!state->context_->ongoing_write_) {
+      state->context_->ongoing_write_ = true;
+      state->context_->DecoupledWriteResponse(state);
+    } else {
+      state->context_->PutTaskBackToQueue(state);
     }
   }
 
-  return !finished;
+  return finished;
+}
+
+bool
+ModelStreamInferHandler::Process(InferHandler::State* state, bool rpc_ok)
+{
+  LOG_VERBOSE(1) << "Process for " << Name() << ", rpc_ok=" << rpc_ok
+                 << ", context " << state->context_->unique_id_ << ", "
+                 << state->unique_id_ << " step " << state->step_;
+
+  // We need an explicit finish indicator. Can't use 'state->step_'
+  // because we launch an async thread that could update 'state's
+  // step_ to be FINISH before this thread exits this function.
+  bool finished = false;
+
+  if (state->step_ == Steps::START) {
+    finished = RequestStartStep(state, rpc_ok);
```

Comment:
Can we pull the step transition from within these helpers to this Process function? It would help in giving a complete picture of the state machine being implemented here.
If not, then may be add a comment describing to which state the state can transition to in these helper functions.

Reply:
This would be nice, but there's a lot of branching in each function where `if (this) state = state1; else state = state2;` It would be non-trivial effort to extract this out I think.